ability to not rollup at index time, make pre aggregation an option

kaijianding opened the pull request:
here is the Proposal https://groups.google.com/forum/#!msg/druid-development/GE8QtSMdoFo/BhEkLIOjBAAJ

In our case, we need to keep all details instead of rollup duplicate rows, so here is the patch.
We believe there should be no impact to query part if not rollup

In this patch
1. we add an "rollup" option in IncrementalIndexSchema and a new NoRollupGranularitySpec
2. add "rollup" option to mergeQueryableIndex() and merge() in IndexMerger. and do not use CombiningIterable in rowMergerFn if not rollup, then all rows will be kept without combine duplicate rows to one row. If not rollup, it is ordered by timestamp instead of timeAndDims
3. changes to related classes and doc due to the 1,2
4. in IncrementalIndex, do not directly use facts map, use FactsHolder instead. if rollup, use RollupFactsHolder which just delegates to the original facts map. if no rollup, use PlainFactsHolder, which keeps all duplicate rows in a list as ConcurrentMap < Long, Queue < Pair < TimeAndDims, Integer > > >, the key is timestamp, the value is the list of < timeAndDims,rowNum > pair.
5. update IncrementalIndexStorageAdapter to use FactsHolder instead of the facts map
6. add rollup info to SegmentMetadataQuery, also in metadata.drd when persist
7. benchmark

vogievetsky commented:
Could you please add the ability to find out if there is rollup in a data source from the segmentMetadata query. This would be very helpful for tools that interface with Druid. Like https://github.com/implydata/pivot and https://github.com/airbnb/caravel

vogievetsky commented:
Could you please add docs for how to use this feature?

vogievetsky commented:
Kickass! Thank you for incorporating my comments!

vogievetsky commented on docs/content/ingestion/index.md:
```diff
@@ -71,6 +72,7 @@ An example dataSchema is shown below:
 | dataSource | String | The name of the ingested datasource. Datasources can be thought of as tables. | yes |
 | parser | JSON Object | Specifies how ingested data can be parsed. | yes |
 | metricsSpec | JSON Object array | A list of [aggregators](../querying/aggregations.html). | yes |
+| rollup | Boolean | rollup or not | no (default == true) |
```
From that I know of the Druid commiter people they will not like the "default to true" settings. Things should always default to false or so they say :-p. Now the 'simple fix' is to change this setting to (say) `disableRollup` and have it default to `false`. **But** I think we should have a conversation about having no rollup be the default in Druid.

'No rollup' is how most DBs out there operate. It would make Druid much easier to evaluate because you would not need to have to think about rollup from day 0.

I would love to get the Druid people @xvrl , @gianm , @cheddar , @fjy , @himanshug , @drcrallen and anyone I am forgetting to comment on possibly making Druid have no rollup by default.

fjy replied:
This will be a fairly large change to have no rollup as default and require updating the version number. FWIW, I think we should keep rollup on by default and re-evaluate for 0.10.0

nishantmonu51 replied:
agree with @fjy,   I am in favor of keeping rollup on by default at least for 0.9

kaijianding commented:
Done, @vogievetsky

nishantmonu51 commented on processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:
```diff
@@ -345,14 +345,23 @@ public static SegmentAnalysis mergeAnalyses(
+    if (arg1.isRollup() != null && arg2.isRollup() != null && !arg1.isRollup() && !arg2.isRollup()) {
+      rollup = Boolean.FALSE;
+    } else {
+      rollup = null;
```
is there any case when this needs to be true instead of null ?

kaijianding replied:
I will modify related code to this logic: when merge SegmentAnalysis, false&false=false, true&true=true, true&false=null

nishantmonu51 commented on processing/src/main/java/io/druid/query/metadata/metadata/SegmentAnalysis.java:
```diff
@@ -38,6 +38,7 @@
   private final long numRows;
   private final Map<String, AggregatorFactory> aggregators;
   private final QueryGranularity queryGranularity;
+  private final Boolean rollup;
```
minor nit: I would prefer primitive types, unless null adds any specific meaning here.

kaijianding replied:
I will modify related code to this logic: when merge SegmentAnalysis, false&false=false, true&true=true, true&false=null
so I need it be Boolean instead of boolean

nishantmonu51 replied:
cool

nishantmonu51 replied:
sounds good.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -200,6 +200,9 @@ public File persist(
         ),
+        // if index is not rollup, then it should be not rollup here
+        // if index is rollup, then it is no need to rollup again
+        false,
```
whether to rollup or not also affects Rowboat ordering, 
In case we want to do rollup, and two intermediate segments are persisted but with different row ordering (based on timestamps only and not considering dimensions),  
I wonder if this could this cause rows to be not rolled up during final merging of intermediate segments ?

kaijianding replied:
In persist stage, if index.rollup==true, persist.rollup==false, then the rowMergerFn(the MergeIterable) actually doesn't change anything(element self and the order) due to having only one underlying iterator, in this case, everything works the same as before.
In segments merge stage, merge.rollup==true, everything should be the same as before due to it is the fallback case before this patch

nishantmonu51 replied:
please modify the comment to reflect that true/false for persist.rollup, does not matter since there is only one iterable and we never really so any reOrdering while merging a single iterable.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -1304,4 +1293,240 @@ public int compare(TimeAndDims lhs, TimeAndDims rhs)
+    }
+
+    @Override
+    public Integer getPriorIndex(TimeAndDims key)
```
this seems a bit hacky, can we move the logic of whether to add a new row or to aggregate to an existing row inside the factsHolder, i.e only have a simple method add(InputRow row) to FactsHolder, and the holder internally takes all the decisions on how to add the row as per its underlying implementation.

kaijianding replied:
To my understanding, the implements of IncrementalIndex.add(InputRow row) in OnheapIncrementalIndex/OffheapIncrementalIndex/OnheapIncrementalIndexBenchmark.MapIncrementalIndex are complicated, it's hard to extract a simple method add(InputRow row) in FactsHolder to add a new row without knowing how to do aggregation. If FactsHolder knows the detail of aggregation, it becomes another similar copy of IncrementalIndex.
So I simply make FactsHolder the replacement of the original facts map, the FactsHolder itself doesn't know when to add a new row/the maxRowCount/how to keep track of rowIndex/clean aggregation environment if lost a race in race condition.

Please correct me if I misunderstood your comment.

nishantmonu51 replied:
How about renaming this method to getRowIndexForRollup(TimeAndDims key) to make it more clear that the underlying impl may not always be a map.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndexSchema.java:
```diff
@@ -67,19 +70,26 @@ public DimensionsSpec getDimensionsSpec()
       this.gran = QueryGranularities.NONE;
       this.dimensionsSpec = new DimensionsSpec(null, null, null);
       this.metrics = new AggregatorFactory[]{};
+      this.rollup = true;
```
minor nit : would be nice If we can extract it as a constant somewhere, DEFAULT_ROLLUP, will be helpful if we plan to change the default in future.

kaijianding replied:
better to keep rollup=true here, because it will affect lots of UT classes if set rollup to false by default.
If we change the default druid rollup behavior to false in future, we only have to modify the DataSchema.java to set rollup=false if rollup option is not set, and leave these intermediate classes unchanged.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/incremental/OffheapIncrementalIndex.java:
```diff
@@ -119,6 +113,7 @@ public OffheapIncrementalIndex(
         new IncrementalIndexSchema.Builder().withMinTimestamp(minTimestamp)
                                             .withQueryGranularity(gran)
                                             .withMetrics(metrics)
+                                            .withRollup(true)
```
use DEFAULT_ROLLUP constant instead

nishantmonu51 commented:
@kaijianding: Not a blocker for this PR, just out of curiousity, did you ran any benchmarks with rollup disabled ? 
would be great to see performance comparison for ingestion, persisting & merging with and without rollup. you can probably use IndexMargeBenchmark & OnheapIncrementalIndexBenchmark.

kaijianding commented:
Will post results here when I done the benchmarks @nishantmonu51

fjy commented:
@kaijianding did you get a chance to get the benchmarks?

fjy commented:
@kaijianding

kaijianding commented:
Will get benchmark by end of this week, @fjy

jon-wei commented:
@kaijianding I looked at the core changes in IncrementalIndex today, seems good so far, will review the rest as well

kaijianding commented:
latest master

Benchmark                           (rowsPerSegment)  (schema)  Mode  Cnt       Score       Error  Units
IncrementalIndexReadBenchmark.read            750000     basic  avgt   25  546006.587 ± 38764.187  us/op
IndexIngestionBenchmark.addRows               75000      basic  avgt   25  710651.752 ± 66651.048  us/op
IndexPersistBenchmark.persist                 75000      basic  avgt   25  2093374.955 ± 146172.738  us/op
IndexPersistBenchmark.persistV9               75000      basic  avgt   25  1757484.706 ±  83394.681  us/op
Benchmark                    (numSegments)  (rowsPerSegment)  (schema)  Mode  Cnt        Score        Error  Units
IndexMergeBenchmark.merge                5             75000     basic  avgt   25  8662055.099 ± 178077.777  us/op
IndexMergeBenchmark.mergeV9              5             75000     basic  avgt   25  7122331.471 ±  83495.704  us/op


rollup with patch

Benchmark                           (rowsPerSegment)  (schema)  Mode  Cnt       Score       Error  Units
IncrementalIndexReadBenchmark.read            750000     basic  avgt   25  579600.132 ± 34073.747  us/op
IndexIngestionBenchmark.addRows               75000      basic  avgt   25  733389.849 ± 77003.987  us/op
IndexPersistBenchmark.persist                 75000      basic  avgt   25  2142651.777 ± 168597.841  us/op
IndexPersistBenchmark.persistV9               75000      basic  avgt   25  1811357.230 ± 127541.039  us/op
Benchmark                    (numSegments)  (rowsPerSegment)  (schema)  Mode  Cnt        Score        Error  Units
IndexMergeBenchmark.merge                5             75000     basic  avgt   25  9109752.827 ± 185592.827  us/op
IndexMergeBenchmark.mergeV9              5             75000     basic  avgt   25  7504100.189 ±  65147.961  us/op

no-rollup with patch

Benchmark                           (rowsPerSegment)  (schema)  Mode  Cnt       Score       Error  Units
IncrementalIndexReadBenchmark.read            750000     basic  avgt   25  603619.117 ± 17928.261  us/op
IndexIngestionBenchmark.addRows               75000      basic  avgt   25  765536.305 ± 47033.539  us/op
IndexPersistBenchmark.persist                 75000      basic  avgt   25  2119441.728 ± 166131.083  us/op
IndexPersistBenchmark.persistV9               75000      basic  avgt   25  1776699.767 ± 111885.336  us/op
Benchmark                    (numSegments)  (rowsPerSegment)  (schema)  Mode  Cnt        Score        Error  Units
IndexMergeBenchmark.merge                5             75000     basic  avgt   25  8491461.225 ± 262106.104  us/op
IndexMergeBenchmark.mergeV9              5             75000     basic  avgt   25  7025990.144 ±  81410.256  us/op


here is the benchmark, @fjy @nishantmonu51

fjy commented:
@kaijianding 

How many iterations did you run? There seems to be a 5-6% slowdown for ingestion, which is currently not acceptable for us. There is also a wide error range. Does running more iterations help reduce this error range? If there is actually a 5-6% slowdown in performance, I think it is worth investigating why the slowdown is occurring as ideally there should be no performance impact when rollup is turned on.

kaijianding commented:
```
master

Benchmark                           (rowsPerSegment)  (schema)  Mode  Cnt       Score       Error  Units
IncrementalIndexReadBenchmark.read            750000     basic  avgt  300  461207.303 ± 1248.121  us/op
IndexIngestionBenchmark.addRows               75000      basic  avgt  200  564279.494 ± 6781.083  us/op
IndexPersistBenchmark.persist                 75000      basic  avgt  100  1490063.885 ± 32357.405  us/op
IndexPersistBenchmark.persistV9               75000      basic  avgt  100  1298518.894 ± 22939.044  us/op
Benchmark                    (numSegments)  (rowsPerSegment)  (schema)  Mode  Cnt        Score        Error  Units
IndexMergeBenchmark.merge                5             75000     basic  avgt  100  6203247.612 ± 47913.348  us/op
IndexMergeBenchmark.mergeV9              5             75000     basic  avgt  100  5243835.387 ± 29471.834  us/op

rollup with patch

Benchmark                           (rowsPerSegment)  (schema)  Mode  Cnt       Score       Error  Units
IncrementalIndexReadBenchmark.read            750000     basic  avgt  300  478140.158 ± 1430.996  us/op
IndexIngestionBenchmark.addRows               75000      basic  avgt  200  560886.776 ± 7041.057  us/op
IndexPersistBenchmark.persist                 75000      basic  avgt  100  1508987.804 ± 34927.774  us/op
IndexPersistBenchmark.persistV9               75000      basic  avgt  100  1301953.483 ± 24832.039  us/op
Benchmark                    (numSegments)  (rowsPerSegment)  (schema)  Mode  Cnt        Score        Error  Units
IndexMergeBenchmark.merge                5             75000     basic  avgt  100  6334795.617 ± 46532.740  us/op
IndexMergeBenchmark.mergeV9              5             75000     basic  avgt  100  5256236.805 ± 31898.007  us/op

no-rollup with patch

Benchmark                           (rowsPerSegment)  (schema)  Mode  Cnt       Score       Error  Units
NoRollupIncrementalIndexReadBenchmark.read            750000     basic  avgt  300  538949.116 ± 1566.411  us/op
NoRollupIndexIngestionBenchmark.addRows               75000      basic  avgt  200  570859.946 ± 3950.365  us/op
NoRollupIndexPersistBenchmark.persist                 75000      basic  avgt  100  1507740.670 ± 32116.040  us/op
NoRollupIndexPersistBenchmark.persistV9               75000      basic  avgt  100  1315232.369 ± 25098.084  us/op
Benchmark                    (numSegments)  (rowsPerSegment)  (schema)  Mode  Cnt        Score        Error  Units
NoRollupIndexMergeBenchmark.merge                5             75000     basic  avgt  100  6055157.853 ± 47004.241  us/op
NoRollupIndexMergeBenchmark.mergeV9              5             75000     basic  avgt  100  5068509.238 ± 32577.070  us/op
```

here is the latest benchmark after running with more iterations on a more powerful machine and with slightly code modification, @fjy

fjy commented:
@kaijianding cool! looks much better

fjy commented on docs/content/ingestion/index.md:
```diff
@@ -173,6 +175,11 @@ handle all formatting decisions on their own, without using the ParseSpec.
 | dimensionExclusions | JSON String array | The names of dimensions to exclude from ingestion. | no (default == [] |
 | spatialDimensions | JSON Object array | An array of [spatial dimensions](../development/geo.html) | no (default == [] |
 
+## Rollup
```
can we move this section and instead have this info in the table that first describes the rollup setting?

fjy replied:
@kaijianding can we remove this section and have this info in the description of rollup in the table above?

fjy commented on processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java:
```diff
@@ -347,14 +347,23 @@ public static SegmentAnalysis mergeAnalyses(
+    if (arg1.isRollup() != null && arg2.isRollup() != null && arg1.isRollup().equals(arg2.isRollup())) {
+      rollup = arg1.isRollup();
+    } else {
+      rollup = null;
```
what does null rollup mean? I feel like whenever we can't make a decision rollup should be set to true with a comment that Druid rolls up data by default

kaijianding replied:
there is a case, if previous segment is rolled up, and the next segment use new spec which is not to rollup, then for segmentMetadataQuery should handle this case and return null if merge=true
Otherwise, rollup should be always true or false

fjy commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -200,6 +200,11 @@ public File persist(
                 indexSpec.getBitmapSerdeFactory().getBitmapFactory()
             )
         ),
+        // if index is not rollup, then it should be not rollup here
```
rolled up not rollup

fjy commented on processing/src/main/java/io/druid/segment/Rowboat.java:
```diff
@@ -126,6 +126,11 @@ public int compareTo(Rowboat rhs)
     return retVal;
   }
 
+  public int compareTimestamp(Rowboat rhs)
```
what purpose does this function add? seems unnecessary

kaijianding replied:
used in IndexMerger.merge when determining rowMergerFn, if rollup=false,only order by timestamp
```
        if (rollup) {
          return CombiningIterable.create(
              new MergeIterable<Rowboat>(
                  Ordering.<Rowboat>natural().nullsFirst(),
                  boats
              ),
              Ordering.<Rowboat>natural().nullsFirst(),
              new RowboatMergeFunction(sortedMetricAggs)
          );
        } else {
          return new MergeIterable<Rowboat>(
              new Ordering<Rowboat>()
              {
                @Override
                public int compare(Rowboat left, Rowboat right)
                {
                  return left.compareTimestamp(right);
                }
              }.nullsFirst(),
              boats
          );
        }
```

fjy replied:
can you just have Longs.compare(left.timestamp, right.timestamp) and avoid creating this function?

kaijianding replied:
sure

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -341,14 +346,15 @@ public int lookupId(String name)
     };
   }
 
+  protected static final boolean DEFAULT_ROLLUP = true;
```
can we define statics at teh top of the file?

fjy commented on server/src/main/java/io/druid/segment/indexing/DataSchema.java:
```diff
@@ -58,6 +59,7 @@ public DataSchema(
       @JsonProperty("parser") Map<String, Object> parser,
       @JsonProperty("metricsSpec") AggregatorFactory[] aggregators,
       @JsonProperty("granularitySpec") GranularitySpec granularitySpec,
+      @JsonProperty("rollup") Boolean rollup,
```
this isn't really a property of the dataschema and should be a part of the TuningConfig

kaijianding replied:
IMO, TuningConfig doesn't essentially change the data itself, only something like partition, format, performance, etc.
But rollup determines if we should keep the detail of the data like we choose how many dimension from columns, so I think it should be part of the data schema

fjy replied:
Okay, I think that is reasonable. Can we please put it in the granularity spec then? The configuration to define the rollup right now (queryGranularity) is there

fjy commented on benchmarks/src/main/java/io/druid/benchmark/indexing/IncrementalIndexReadBenchmark.java:
```diff
@@ -106,13 +106,18 @@ public void setup() throws IOException
 
   }
 
+  protected boolean isRollup() {
```
what's the purpose of this function?

fjy replied:
instead of having this as a function, can we parameterize the benchmark similar to what we do for tests?

fjy replied:
see OnheapIncrementalIndexBenchmark.java for an example

kaijianding replied:
sure

fjy commented on benchmarks/src/main/java/io/druid/benchmark/indexing/IndexIngestionBenchmark.java:
```diff
@@ -100,13 +100,18 @@ public void setup2() throws IOException
     incIndex = makeIncIndex();
   }
 
+  protected boolean isRollup() {
```
same comment, can we parameterize things?

fjy commented on benchmarks/src/main/java/io/druid/benchmark/indexing/IndexMergeBenchmark.java:
```diff
@@ -148,13 +148,18 @@ public void setup() throws IOException
     }
   }
 
+  protected boolean isRollup() {
```
same comment about parameterization

fjy commented on benchmarks/src/main/java/io/druid/benchmark/indexing/IndexPersistBenchmark.java:
```diff
@@ -149,13 +149,18 @@ public void teardown() throws IOException
     incIndex = null;
   }
 
+  protected boolean isRollup() {
```
same comment as above

fjy commented on benchmarks/src/main/java/io/druid/benchmark/indexing/NoRollupIncrementalIndexReadBenchmark.java:
```diff
@@ -0,0 +1,39 @@
+@Fork(jvmArgsPrepend = "-server", value = 1)
+@Warmup(iterations = 10)
+@Measurement(iterations = 25)
+public class NoRollupIncrementalIndexReadBenchmark extends IncrementalIndexReadBenchmark
```
i don't think this is needed

fjy commented on benchmarks/src/main/java/io/druid/benchmark/indexing/NoRollupIndexIngestionBenchmark.java:
```diff
@@ -0,0 +1,44 @@
+@Fork(jvmArgsPrepend = "-server", value = 1)
+@Warmup(iterations = 10)
+@Measurement(iterations = 25)
+public class NoRollupIndexIngestionBenchmark extends IndexIngestionBenchmark
```
is this class needed?

fjy commented on benchmarks/src/main/java/io/druid/benchmark/indexing/NoRollupIndexMergeBenchmark.java:
```diff
@@ -0,0 +1,39 @@
+@Fork(jvmArgsPrepend = "-server", value = 1)
+@Warmup(iterations = 10)
+@Measurement(iterations = 25)
+public class NoRollupIndexMergeBenchmark extends IndexMergeBenchmark
```
is this class needed?

fjy commented on benchmarks/src/main/java/io/druid/benchmark/indexing/NoRollupIndexPersistBenchmark.java:
```diff
@@ -0,0 +1,39 @@
+@Fork(jvmArgsPrepend = "-server", value = 1)
+@Warmup(iterations = 10)
+@Measurement(iterations = 25)
+public class NoRollupIndexPersistBenchmark extends IndexPersistBenchmark
```
is this class needed?

fjy commented on docs/content/ingestion/tasks.md:
```diff
@@ -146,14 +146,18 @@ Append tasks append a list of segments together into a single segment (one after
 
-Merge tasks merge a list of segments together. Any common timestamps are merged. The grammar is:
+Merge tasks merge a list of segments together. Any common timestamps are merged.
+If rollup==false, number of rows are not reduced and rows are reordered by timestamp.
```
If rollup is disabled as part of ingestion, common timestamps are not merged and rows are reorderded by their timestamp.

fjy commented on docs/content/querying/segmentmetadataquery.md:
```diff
@@ -11,6 +11,7 @@ Segment metadata queries return per-segment information about:
 * Interval the segment covers
 * Column type of all the columns in the segment
 * Estimated total segment byte size in if it was stored in a flat format
+* Is segment rollup when persist
```
Is the segment rolled up

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java:
```diff
@@ -515,11 +516,11 @@ protected File mergeQueryableIndex(
       if (config.isBuildV9Directly()) {
         return HadoopDruidIndexerConfig.INDEX_MERGER_V9.mergeQueryableIndex(
-            indexes, aggs, file, config.getIndexSpec(), progressIndicator
+            indexes, config.getSchema().getDataSchema().isRollup(), aggs, file, config.getIndexSpec(), progressIndicator
```
can rollup be a part of the indexSpec?

fjy replied:
or just not in data schema

kaijianding commented on docs/content/querying/segmentmetadataquery.md:
```diff
@@ -139,6 +140,11 @@ null if the aggregators are unknown or unmergeable (if merging is enabled).
 
+#### rollup
+
+* `rollup` in the result is true/false/null.
```
there is a case, if previous segment is rolled up, and the next segment use new spec which is not to rollup, then for segmentMetadataQuery should handle this case and return null if merge=true
Otherwise, rollup should be always true or false

fjy commented on docs/content/ingestion/tasks.md:
```diff
@@ -146,14 +146,18 @@ Append tasks append a list of segments together into a single segment (one after
     "id": <task_id>,
     "dataSource": <task_datasource>,
     "aggregations": <list of aggregators>,
+    "rollup": <rollup when merge>,
```
plz change to "whether or not to rollup data during a merge"

kaijianding commented:
I moved the rollup option from DataSchema to the new NoRollupGranularitySpec whose queryGranularity is always none, and updated the ingestion/index.md

fjy commented:
@kaijianding I don't think we need a new type of granularity spec, I think the default (uniform granularity spec) should have the option of turning rollup off right now. The uniform granularity spec currently has the parameter queryGranularity which controls how data gets rolled up.

fjy commented on indexing-service/src/main/java/io/druid/indexing/common/task/MergeTask.java:
```diff
@@ -52,12 +53,14 @@ public MergeTask(
       @JsonProperty("segments") List<DataSegment> segments,
       @JsonProperty("aggregations") List<AggregatorFactory> aggregators,
+      @JsonProperty("rollup") Boolean rollup,
       @JsonProperty("indexSpec") IndexSpec indexSpec,
```
The documentation for the mergeTask needs to be updated

kaijianding replied:
already updated in tasks.md

fjy commented on processing/src/main/java/io/druid/segment/incremental/OffheapIncrementalIndex.java:
```diff
@@ -114,6 +108,7 @@ public OffheapIncrementalIndex(
         new IncrementalIndexSchema.Builder().withMinTimestamp(minTimestamp)
                                             .withQueryGranularity(gran)
                                             .withMetrics(metrics)
+                                            .withRollup(DEFAULT_ROLLUP)
```
can this use incrementalIndexSchema.isRollup?

kaijianding replied:
will change to Incrementalindexschema. DEFAULT_ROLLUP

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -73,15 +74,20 @@
  */
 public abstract class IncrementalIndex<AggregatorType> implements Iterable<Row>, Closeable
 {
+  protected static final boolean DEFAULT_ROLLUP = true;
```
can we put this in incrementalindexschema?

kaijianding replied:
sure

kaijianding commented:
I modified the code according to your comments, and rebased it, do you have more comments? @fjy

fjy commented:
@kaijianding can you fix the merge conflicts?

kaijianding commented:
fixed. @fjy

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -1320,4 +1308,288 @@ public int compare(TimeAndDims lhs, TimeAndDims rhs)
+    }
+
+    @Override
+    public Integer getRowIndexForRollup(TimeAndDims key)
```
can we call this getPriorIndex()?

fjy commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -1320,4 +1308,288 @@ public int compare(TimeAndDims lhs, TimeAndDims rhs)
+     * @return the previous value associated with the specified key, or
+     * {@code null} if there was no mapping for the key.
+     */
+    Integer getRowIndexForRollup(TimeAndDims key);
```
can we call this getPriorIndex()?

fjy commented:
@kaijianding mostly looks good, I have a few more minor comments

@jon-wei @nishantmonu51  can u guys take a look ?

kaijianding commented:
renamed. @fjy

jon-wei commented:
@fjy sure I'll take a look at this today

jon-wei commented on docs/content/ingestion/tasks.md:
```diff
@@ -159,14 +159,18 @@ Append tasks append a list of segments together into a single segment (one after
 
-Merge tasks merge a list of segments together. Any common timestamps are merged. The grammar is:
+Merge tasks merge a list of segments together. Any common timestamps are merged.
+If rollup is disabled as part of ingestion, common timestamps are not merged and rows are reorderded by their timestamp.
```
"reorderded" should be "reordered"

jon-wei commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -409,14 +420,28 @@ public String apply(@Nullable String input)
+                @Override
+                public int compare(Rowboat left, Rowboat right)
+                {
+                  return Longs.compare(left.getTimestamp(), right.getTimestamp());
```
Should this use the Rowboat's compareTo() function instead?

kaijianding replied:
For no-rollup, the inner structure is Map<timestamp, List<timeAndDims, rowNum>>. When timestamp is same, the order of the list is the order the item added, it is not ordered by timeAndDims.
So it is not able to make the ordering well-defined even using Rowboat's compareTo() function in the merge sort process.

jon-wei replied:
If a well-defined ordering is really needed, maybe the ConcurrentLinkedQueues in the PlainFactsHolder map could be changed to priority queues,  but this seems fine for now if there are no problems with queries.

Can you add the new no-rollup indexes from `TestIndex` to the `makeQueryRunners` in `QueryRunnerTestHelper` and check that all of the query unit tests pass?

kaijianding replied:
added. @jon-wei 
To pass SelectQueryRunnerTest, I changed ConcurrentLinkedQueue to ConcurrentLinkedDeque to use queue.descendingIterator()

jon-wei commented:
looks good to me aside from my last two comments

fjy commented:
👍

fjy commented:
@kaijianding can you run benchmarks again just to make sure that nothing has been impacted with all the changes during code review?

jon-wei commented:
:+1: if new benchmarks look good

kaijianding commented:
```
master

Benchmark                                      (rowsPerSegment)  (schema)  Mode  Cnt       Score      Error  Units
IncrementalIndexReadBenchmark.read                     750000    basic  avgt  500  385330.680 ±  891.106  us/op
IncrementalIndexReadBenchmark.readWithFilters          750000    basic  avgt  500  502956.956 ± 1054.732  us/op
IndexIngestionBenchmark.addRows                        75000     basic  avgt  200  501676.186 ± 2953.806  us/op
IndexPersistBenchmark.persist                          75000     basic  avgt  100  1346216.134 ± 14633.695  us/op
IndexPersistBenchmark.persistV9                        75000     basic  avgt  100  1179403.111 ± 12030.273  us/op
Benchmark                    (numSegments)     (rowsPerSegment)  (schema)  Mode  Cnt       Score      Error  Units
IndexMergeBenchmark.merge                5             75000     basic  avgt  100  5835544.731 ± 22592.354  us/op
IndexMergeBenchmark.mergeV9              5             75000     basic  avgt  100  5033092.133 ± 13562.767  us/op

rollup with patch

Benchmark                                   (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt       Score      Error  Units
IncrementalIndexReadBenchmark.read              true            750000     basic  avgt  500  379315.134 ± 885.494  us/op
IncrementalIndexReadBenchmark.readWithFilters   true            750000     basic  avgt  500  400021.287 ± 843.226  us/op
IndexIngestionBenchmark.addRows                 true             75000     basic  avgt  100  509449.271 ± 5378.047  us/op
IndexPersistBenchmark.persist                   true             75000     basic  avgt  100  1329498.943 ± 14548.436  us/op
IndexPersistBenchmark.persistV9                 true             75000     basic  avgt  100  1182184.226 ± 11937.399  us/op
Benchmark                    (numSegments)  (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt        Score       Error  Units
IndexMergeBenchmark.merge                5      true             75000     basic  avgt  100  5739894.613 ± 15663.028  us/op
IndexMergeBenchmark.mergeV9              5      true             75000     basic  avgt  100  5019952.001 ± 10244.254  us/op

norollup

IncrementalIndexReadBenchmark.read             false            750000     basic  avgt  500  459174.614 ± 642.678  us/op
IncrementalIndexReadBenchmark.readWithFilters  false            750000     basic  avgt  500  465450.957 ± 452.000  us/op
IndexIngestionBenchmark.addRows                false             75000     basic  avgt  100  618939.096 ± 17863.195  us/op
IndexPersistBenchmark.persist                  false             75000     basic  avgt  100  1330675.790 ± 13748.631  us/op
IndexPersistBenchmark.persistV9                false             75000     basic  avgt  100  1176238.798 ± 12016.404  us/op
IndexMergeBenchmark.merge                5     false             75000     basic  avgt  100  5636242.340 ± 18749.393  us/op
IndexMergeBenchmark.mergeV9              5     false             75000     basic  avgt  100  4929684.534 ± 16980.737  us/op
```

here is the benchmark. @fjy @jon-wei 
Basically, there is no change in rollup scenario. no rollup scenario is 20% slower in read and addRows.

BTW, I noticed that the jmh framework would call each benchmark method 1~2 times each iteration. In this way, the second call, it is not called in clean environment. Fg, IndexIngestionBenchmark.addRows test, the second call will cause incIndex to add more rows than expected.
